### PR TITLE
[zero]support zero2 with gradient accumulation

### LINKED ISSUE #4510: [FEATURE]: Support gradient accumulation with ZeRO2
### Describe the feature

As we know, ZeRO2 would split the gradient, which is incompatible with gradient accumulation.
However, they are not that incompatible.
For instance, we can accumulate the gradients belonging to each rank after communication.

Drawback:
This version of gradient accumulation saves no communication.

Advantage:
For some users who need a large batch to train their model(for model convergence, e.g.) but have limited GPU memory, gradient accumulation can be a solution.

### PR DESCRIPTION
**Pull request by Gy-Lu**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

fixed #4510 

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

This PR support gradient accumulation when using zero2

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

**Review by kurisusnowdeng — COMMENTED** [review]
BTW, shall we rename zero1/2 as `ZeroRedundancyOptimizer` from torch, just for purpose of avoiding creating new names or concepts

**Comment by Gy-Lu** [comment]
> BTW, shall we rename zero1/2 as `ZeroRedundancyOptimizer` from torch, just for purpose of avoiding creating new names or concepts

Make sense.
However, if we rename it as `ZeroRedundancyOptimizer`, it should contain ZeRO3 as well.

We may figure out it with a new PR.

**Comment by kurisusnowdeng** [comment]
> > BTW, shall we rename zero1/2 as `ZeroRedundancyOptimizer` from torch, just for purpose of avoiding creating new names or concepts
> 
> Make sense. However, if we rename it as `ZeroRedundancyOptimizer`, it should contain ZeRO3 as well.
> 
> We may figure out it with a new PR.

Not necessary. FSDP is used for zero3 in torch. Likewise, we have Gemini

**Review by kurisusnowdeng — APPROVED** [review]

**Comment by Gy-Lu** [comment]
> > > BTW, shall we rename zero1/2 as `ZeroRedundancyOptimizer` from torch, just for purpose of avoiding creating new names or concepts
> > 
> > 
> > Make sense. However, if we rename it as `ZeroRedundancyOptimizer`, it should contain ZeRO3 as well.
> > We may figure out it with a new PR.
> 
> Not necessary. FSDP is used for zero3 in torch. Likewise, we have Gemini

OK
